internal/pkg/cli/task_run.go

```diff
@@ -134,6 +176,155 @@ func (o *runTaskOpts) Ask() error {
 	return nil
 }
 
+// Execute create or update task resources and run the task.
+func (o *runTaskOpts) Execute() error {
+	if o.image == "" && o.dockerfilePath == "" {
+		o.dockerfilePath = defaultDockerfilePath
+	}
+
+	if o.imageTag == "" {
+		o.imageTag = defaultImageTag
+	}
+
+	if err := o.getNetworkConfig(); err != nil {
+		return err
+	}
+
+	cluster, err := o.getCluster()
+	if err != nil {
+		return err
+	}
+
+	if err := o.createAndUpdateTaskResources(); err != nil {
+		return err
+	}
+
+	//if image is not provided, then we build the image and push to ECR repo
+	if o.image == "" {
+		uri, err := o.pushToECRRepo()
+		if err != nil {
+			return err
+		}
+		o.image = fmt.Sprintf(fmtImageURL, uri, o.imageTag)
+
+		// update image to stack
+		if err := o.createAndUpdateTaskResources(); err != nil {
+			return err
+		}
+	}
+
+	return o.runTask(cluster)
+}
+
+func (o *runTaskOpts) getCluster() (string, error) {
+	if o.env == config.EnvNameNone {
+		clusters, err := o.ecsGetter.DefaultClusters()
+		if err != nil {
+			return "", fmt.Errorf("get default cluster: %w", err)
+		}
+		return clusters[0], nil
+	}
+
+	clusters, err := o.resourceGetter.GetResourcesByTags(clusterResourceType, map[string]string{
+		deploy.AppTagKey: o.AppName(),
+		deploy.EnvTagKey: o.env,
+	})
+
+	if err != nil {
+		return "", fmt.Errorf("get cluster: %w", err)
+	}
+
+	return clusters[0], nil
+}
+
+func (o *runTaskOpts) runTask(cluster string) error {
+	o.spinner.Start(fmt.Sprintf("Waiting for %s to be running.", o.groupName))
+	err := o.starter.RunTask(ecs.RunTaskInput{
+		Cluster:        cluster,
+		Count:          o.count,
+		Subnets:        o.subnets,
+		SecurityGroups: o.securityGroups,
+		TaskFamilyName: fmt.Sprintf(fmtTaskFamilyName, o.groupName),
+	})
+
+	if err != nil {
+		o.spinner.Stop(log.Serrorln(fmt.Sprintf("Failed to run %s", o.groupName)))
+		return fmt.Errorf("run task %s: %w", o.groupName, err)
+	}
+
+	o.spinner.Stop(log.Ssuccessln(fmt.Sprintf("Task(s) %s is running.", o.groupName)))
+	return nil
+}
+
+func (o *runTaskOpts) getNetworkConfig() error {
+	if o.subnets == nil {
+		subnets, err := o.vpcGetter.GetSubnetIDs(o.AppName(), o.env)
+		if err != nil {
+			return fmt.Errorf("get subnet IDs from environment %s: %w", o.env, err)
+		}
+		o.subnets = subnets
+	}
+
+	// if the task is to be run in None environment (default environment), leave security groups as nil since it is not
+	// a required field for run task
+	if o.securityGroups == nil {
+		securityGroups, err := o.vpcGetter.GetSecurityGroups(o.AppName(), o.env)
+		if err != nil {
+			return fmt.Errorf("get security groups from environment %s: %w", o.env, err)
+		}
+		o.securityGroups = securityGroups
+	}
+	return nil
+}
+
+func (o *runTaskOpts) createAndUpdateTaskResources() error {
+	o.spinner.Start(fmt.Sprintf("Deploying resources for task %s", color.HighlightUserInput(o.groupName)))
```

Comment:
i wonder if the ux could be confusing because of the spinner message. Suppose we use local image then the spinner will show up first with message saying "Deploying resources for task..." then comes another spinner saying the same thing. I would recommend making them separate methods with `createTaskResources` and spinner message "creating task resources ...." and `updateTaskResources" and spinner message like "Updating task resources with image name...", which will be more intuitive to users.

Reply:
yes that makes sense. What do you think about using `deploying` instead of `creating` in the first call? Because it is not necessarily creating resources, it could be updating the same stack that's been created in a previous run of `task run` command.

Reply:
Yeah "deploying" sounds better.